DeviceBridge/Services/ConnectionManager.cs

```diff
@@ -186,25 +187,18 @@ public async Task AssertDeviceConnectionOpenAsync(string deviceId, bool temporar
                     _hasPermanentConnection.AddOrUpdate(deviceId, true, (key, oldValue) => true);
                 }
 
-                // If desired, force the disposal of the current client if it is in a permanent failure state.
-                if (recreateFailedClient)
+                // Dispose the current client if it is in a permanent failure state.
+                if (_clientStatuses.TryGetValue(deviceId, out (ConnectionStatus status, ConnectionStatusChangeReason reason) currentStatus))
                 {
-                    if (_clientStatuses.TryGetValue(deviceId, out (ConnectionStatus status, ConnectionStatusChangeReason reason) currentStatus))
+                    // Permanent failure state, taken from https://github.com/Azure-Samples/azure-iot-samples-csharp/tree/master/iot-hub/Samples/device/DeviceReconnectionSample
+                    bool isFailed = currentStatus.status == ConnectionStatus.Disconnected;
+
+                    if (isFailed && _clients.TryRemove(deviceId, out DeviceClient existingClient))
                     {
-                        // Permanent failure states taken from https://github.com/Azure-Samples/azure-iot-samples-csharp/tree/master/iot-hub/Samples/device/DeviceReconnectionSample
-                        bool isFailed = currentStatus.status == ConnectionStatus.Disconnected &&
-                            (currentStatus.reason == ConnectionStatusChangeReason.Device_Disabled ||
-                            currentStatus.reason == ConnectionStatusChangeReason.Bad_Credential ||
-                            currentStatus.reason == ConnectionStatusChangeReason.Communication_Error ||
-                            currentStatus.reason == ConnectionStatusChangeReason.Retry_Expired);
-
-                        if (isFailed && _clients.TryRemove(deviceId, out DeviceClient existingClient))
-                        {
-                            _logger.Info("Disposing existing failed client for device {deviceId}", deviceId);
-                            await existingClient.CloseAsync();
-                            existingClient.Dispose();
-                            existingClient.SetConnectionStatusChangesHandler(null);
-                        }
+                        _logger.Info("Disposing existing failed client for device {deviceId}", deviceId);
+                        await existingClient.CloseAsync();
```

Comment:
Shall we do null check before call close and dispose?

Reply:
No need. We're pulling the `existingClient` from the client dictionary, so it's never null.

Reply:
_clients.TryRemove(deviceId, out DeviceClient existingClient), will this get a null client out if deviceid not existed in dictionary?

Reply:
No, `TryRemove` returns false if the key doesn't exist, so we'll not execute this code (the `if` is only executed if the client exists)